include/oneapi/dpl/pstl/histogram_impl.h

```diff
@@ -32,16 +35,86 @@ namespace dpl
+    using _HistogramValueT = typename std::iterator_traits<_RandomAccessIterator>::value_type;
+    static_assert(oneapi::dpl::__internal::__is_serial_tag_v<_Tag> ||
+                  oneapi::dpl::__internal::__is_parallel_forward_tag_v<_Tag>);
+    __pattern_fill(_Tag{}, std::forward<_ExecutionPolicy>(__exec), __histogram_first, __histogram_first + __num_bins,
+                   _HistogramValueT{0});
+    __brick_histogram(__first, __last, __func, __histogram_first, typename _Tag::__is_vector{});
+}
+
+template <class _IsVector, class _ExecutionPolicy, class _RandomAccessIterator1, class _Size, class _IdxHashFunc,
+          class _RandomAccessIterator2>
+void
+__pattern_histogram(oneapi::dpl::__internal::__parallel_tag<_IsVector>, _ExecutionPolicy&& __exec,
+                    _RandomAccessIterator1 __first, _RandomAccessIterator1 __last, _Size __num_bins,
+                    _IdxHashFunc __func, _RandomAccessIterator2 __histogram_first)
+{
+    using __backend_tag = typename oneapi::dpl::__internal::__parallel_tag<_IsVector>::__backend_tag;
+    using _HistogramValueT = typename std::iterator_traits<_RandomAccessIterator2>::value_type;
+    using _DiffType = typename std::iterator_traits<_RandomAccessIterator2>::difference_type;
+
+    _DiffType __n = __last - __first;
+    if (__n == 0)
+    {
+        // when n == 0, we must fill the output histogram with zeros
+        __pattern_fill(oneapi::dpl::__internal::__parallel_tag<_IsVector>{}, std::forward<_ExecutionPolicy>(__exec),
+                       __histogram_first, __histogram_first + __num_bins, _HistogramValueT{0});
+    }
+    else
+    {
+        auto __tls = __par_backend::__make_etls<std::vector<_HistogramValueT>>(__num_bins, _HistogramValueT{0});
 
-    static_assert(sizeof(_Size) == 0 /*false*/,
-                  "Histogram API is currently unsupported for policies other than device execution policies");
+        //main histogram loop
+        //TODO: add defaulted grain-size option for __parallel_for and use larger one here to account for overhead
+        __par_backend::__parallel_for(
+            __backend_tag{}, __exec, __first, __last,
+            [__func, &__tls](_RandomAccessIterator1 __first_local, _RandomAccessIterator1 __last_local) {
+                __internal::__brick_histogram(__first_local, __last_local, __func,
+                                              __tls.get_for_current_thread().begin(), _IsVector{});
+            });
+        // now accumulate temporary storage into output global histogram
+        const std::size_t __num_temporary_copies = __tls.size();
+        __par_backend::__parallel_for(
+            __backend_tag{}, std::forward<_ExecutionPolicy>(__exec), __histogram_first, __histogram_first + __num_bins,
+            [__num_temporary_copies, __histogram_first, &__tls](auto __global_histogram_first,
+                                                                auto __global_histogram_last) {
```

Comment:
The parameter names are confusing, specifically the `__global` prefix. Semantically, the parameters are "local first" and "local last" I believe,
I understand that you mean the histogram is global (as opposed to per-thread), but nevertheless it's confusing.

Reply:
Refactored with the indices, and renamed to `__hist_start_id` and `__hist_end_id`.